File: src/utils/validation.ts
/** Pure client-side validation. No DOM, no DB — safe to unit test directly. */

import type { CustomerInput } from '../types';
import { parseAmountToMinor, validateAmount } from './money';
import { localInputToIso } from './date';

export type FieldErrors<T extends string = string> = Partial<Record<T, string>>;

export type CustomerField =
  | 'name'
  | 'address'
  | 'phone'
  | 'room_number'
  | 'number_of_persons'
  | 'id_front'
  | 'id_back'
  | 'amount'
  | 'check_in_date'
  | 'check_out_date';

export const LIMITS = {
  nameMin: 2,
  nameMax: 80,
  addressMin: 5,
  addressMax: 250,
  phoneMin: 7,
  phoneMax: 15,
  personsMin: 1,
  personsMax: 20,
} as const;

/** Keeps digits only, so `+91 98470-12345` counts as 12 digits. */
export function phoneDigits(phone: string): string {
  return phone.replace(/\D/g, '');
}

export function validateName(value: string): string | undefined {
  const v = value.trim();
  if (!v) return 'Customer name is required';
  if (v.length < LIMITS.nameMin) return 'Name looks too short';
  if (v.length > LIMITS.nameMax) return `Name must be under ${LIMITS.nameMax} characters`;
  return undefined;
}

export function validateAddress(value: string): string | undefined {
  const v = value.trim();
  if (!v) return 'Full address is required';
  if (v.length < LIMITS.addressMin) return 'Please enter the full address';
  if (v.length > LIMITS.addressMax) return `Address must be under ${LIMITS.addressMax} characters`;
  return undefined;
}

export function validatePhone(value: string): string | undefined {
  const v = value.trim();
  if (!v) return 'Phone number is required';
  if (!/^[+]?[\d\s\-()]+$/.test(v)) return 'Phone number contains invalid characters';
  const digits = phoneDigits(v);
  if (digits.length < LIMITS.phoneMin) return 'Phone number is too short';
  if (digits.length > LIMITS.phoneMax) return 'Phone number is too long';
  return undefined;
}

export function validateRoomNumber(value: string): string | undefined {
  if (!value.trim()) return 'Please select a room';
  return undefined;
}

export function validatePersons(value: number | string): string | undefined {
  const raw = typeof value === 'string' ? value.trim() : value;
  if (raw === '' || raw === null || raw === undefined) return 'Number of persons is required';
  const n = Number(raw);
  if (!Number.isFinite(n)) return 'Enter a valid number';
  if (!Number.isInteger(n)) return 'Number of persons must be a whole number';
  if (n < LIMITS.personsMin) return 'At least 1 person is required';
  if (n > LIMITS.personsMax) return `Maximum ${LIMITS.personsMax} persons per room`;
  return undefined;
}

/** Room labels are used as file-safe identifiers and must stay short. */
export function validateNewRoomNumber(
  value: string,
  existing: readonly string[],
): string | undefined {
  const v = value.trim();
  if (!v) return 'Room number is required';
  if (v.length > 10) return 'Room number must be 10 characters or fewer';
  if (!/^[A-Za-z0-9\-/ ]+$/.test(v)) return 'Use letters, numbers, spaces, - or / only';
  if (existing.some((r) => r.toLowerCase() === v.toLowerCase())) return 'That room already exists';
  return undefined;
}

/** How far ahead a hand-entered time may be, allowing for clock drift. */
export const MAX_FUTURE_HOURS = 24;

/** A time typed by staff: blank means "use the default". */
export function validateCheckInTime(value: string, now: Date = new Date()): string | undefined {
  if (!value.trim()) return 'Check-in date and time is required';
  const iso = localInputToIso(value);
  if (!iso) return 'Enter a valid date and time';
  const when = new Date(iso).getTime();
  if (when - now.getTime() > MAX_FUTURE_HOURS * 3_600_000) {
    return 'Check-in cannot be more than a day in the future';
  }
  if (new Date(iso).getFullYear() < 2000) return 'That date looks too far in the past';
  return undefined;
}

export function validateCheckOutTime(
  value: string,
  checkInValue: string,
  now: Date = new Date(),
): string | undefined {
  if (!value.trim()) return undefined; // still staying
  const iso = localInputToIso(value);
  if (!iso) return 'Enter a valid date and time';
  if (new Date(iso).getTime() - now.getTime() > MAX_FUTURE_HOURS * 3_600_000) {
    return 'Check-out cannot be more than a day in the future';
  }
  const checkInIso = localInputToIso(checkInValue);
  if (checkInIso && new Date(iso).getTime() < new Date(checkInIso).getTime()) {
    return 'Check-out cannot be before check-in';
  }
  return undefined;
}

export interface CustomerFormValues
  extends Omit<CustomerInput, 'number_of_persons' | 'amount_minor' | 'check_in_date' | 'check_out_date'> {
  number_of_persons: string | number;
  /** Raw text from the amount field; blank means "not recorded". */
  amount: string;
  /** `datetime-local` values (YYYY-MM-DDTHH:mm), local time. */
  check_in_at: string;
  check_out_at: string;
}

/**
 * Validates the whole check-in form. `hasIdFront` / `hasIdBack` describe
 * whether an image is present (either newly captured or already stored).
 */
export function validateCustomerForm(
  values: CustomerFormValues,
  opts: { hasIdFront: boolean; hasIdBack: boolean },
): FieldErrors<CustomerField> {
  const errors: FieldErrors<CustomerField> = {};
  const set = (field: CustomerField, message: string | undefined) => {
    if (message) errors[field] = message;
  };

  set('name', validateName(values.name));
  set('address', validateAddress(values.address));
  set('phone', validatePhone(values.phone));
  set('room_number', validateRoomNumber(values.room_number));
  set('number_of_persons', validatePersons(values.number_of_persons));
  set('amount', validateAmount(values.amount));
  set('check_in_date', validateCheckInTime(values.check_in_at));
  set('check_out_date', validateCheckOutTime(values.check_out_at, values.check_in_at));
  if (!opts.hasIdFront) set('id_front', 'ID proof front photo is required');
  if (!opts.hasIdBack) set('id_back', 'ID proof back photo is required');

  return errors;
}

export function hasErrors(errors: FieldErrors): boolean {
  return Object.values(errors).some(Boolean);
}

/** Normalises form values into the shape the service layer expects. */
export function normaliseCustomerInput(values: CustomerFormValues): CustomerInput {
  const amount = parseAmountToMinor(values.amount);
  const checkIn = localInputToIso(values.check_in_at);
  const checkOut = localInputToIso(values.check_out_at);
  return {
    name: values.name.trim().replace(/\s+/g, ' '),
    address: values.address.trim(),
    phone: values.phone.trim(),
    room_number: values.room_number.trim(),
    number_of_persons: Number(values.number_of_persons),
    // validateCustomerForm rejects unparseable input before this runs.
    amount_minor: amount === undefined ? null : amount,
    ...(checkIn ? { check_in_date: checkIn } : {}),
    ...(checkOut ? { check_out_date: checkOut } : {}),
  };
}

/* ------------------------------------------------------------------ */
/* Additional guests                                                   */
/* ------------------------------------------------------------------ */

export type GuestField = 'name' | 'phone';

/** The name is required once you start recording someone. */
export function validateGuestName(value: string): string | undefined {
  const v = value.trim();
  if (!v) return 'Guest name is required';
  if (v.length < LIMITS.nameMin) return 'Name looks too short';
  if (v.length > LIMITS.nameMax) return `Name must be under ${LIMITS.nameMax} characters`;
  return undefined;
}

/** Optional: a companion often has no separate number on file. */
export function validateGuestPhone(value: string): string | undefined {
  if (!value.trim()) return undefined;
  return validatePhone(value);
}

export function validateGuestForm(values: {
  name: string;
  phone: string;
}): FieldErrors<GuestField> {
  const errors: FieldErrors<GuestField> = {};
  const name = validateGuestName(values.name);
  if (name) errors.name = name;
  const phone = validateGuestPhone(values.phone);
  if (phone) errors.phone = phone;
  return errors;
}

export function normaliseGuestInput(values: { name: string; phone: string }): {
  name: string;
  phone: string | null;
} {
  const phone = values.phone.trim();
  return {
    name: values.name.trim().replace(/\s+/g, ' '),
    phone: phone === '' ? null : phone,
  };
}

export function validatePin(pin: string): string | undefined {
  if (!/^\d{4,8}$/.test(pin)) return 'PIN must be 4 to 8 digits';
  return undefined;
}

File: lib/utils/validation.dart
/// Form-field validation rules — mirrors `src/utils/validation.ts`. Dates are
/// held as real `DateTime` values here (there is no HTML
/// `datetime-local` string to bridge through, unlike the original web app).
library;

import 'money.dart';

class Limits {
  static const nameMin = 2;
  static const nameMax = 80;
  static const addressMin = 5;
  static const addressMax = 250;
  static const phoneMin = 7;
  static const phoneMax = 15;
  static const personsMin = 1;
  static const personsMax = 20;
}

/// Clock-drift allowance: a hand-entered check-in/out time can be at most
/// this many hours in the future.
const maxFutureHours = 24;

String _phoneDigits(String phone) => phone.replaceAll(RegExp(r'\D'), '');

String? validateName(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) return 'Name is required.';
  if (trimmed.length < Limits.nameMin || trimmed.length > Limits.nameMax) {
    return 'Name must be ${Limits.nameMin}-${Limits.nameMax} characters.';
  }
  return null;
}

String? validateAddress(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) return 'Address is required.';
  if (trimmed.length < Limits.addressMin || trimmed.length > Limits.addressMax) {
    return 'Address must be ${Limits.addressMin}-${Limits.addressMax} characters.';
  }
  return null;
}

final _phoneCharPattern = RegExp(r'^[+]?[\d\s\-()]+$');

String? validatePhone(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) return 'Phone number is required.';
  if (!_phoneCharPattern.hasMatch(trimmed)) return 'Phone number has invalid characters.';
  final digits = _phoneDigits(trimmed).length;
  if (digits < Limits.phoneMin || digits > Limits.phoneMax) {
    return 'Phone number must have ${Limits.phoneMin}-${Limits.phoneMax} digits.';
  }
  return null;
}

String? validateRoomNumber(String value) {
  if (value.trim().isEmpty) return 'Room is required.';
  return null;
}

String? validatePersons(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) return 'Number of persons is required.';
  final asNum = num.tryParse(trimmed);
  if (asNum == null || !asNum.isFinite) return 'Enter a valid number.';
  if (asNum != asNum.roundToDouble()) return 'Number of persons must be a whole number.';
  final persons = asNum.toInt();
  if (persons < Limits.personsMin || persons > Limits.personsMax) {
    return 'Number of persons must be ${Limits.personsMin}-${Limits.personsMax}.';
  }
  return null;
}

final _roomNumberPattern = RegExp(r'^[A-Za-z0-9\-/ ]+$');

/// Validates a *new* room number: length, charset (room labels double as
/// filesystem path segments), and no case-insensitive duplicate.
String? validateNewRoomNumber(String value, List<String> existing) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) return 'Room number is required.';
  if (trimmed.length > 10) return 'Room number must be at most 10 characters.';
  if (!_roomNumberPattern.hasMatch(trimmed)) {
    return 'Room number can only contain letters, numbers, spaces, - and /.';
  }
  final lower = trimmed.toLowerCase();
  if (existing.any((r) => r.toLowerCase() == lower)) {
    return 'Room $trimmed already exists.';
  }
  return null;
}

String? validateCheckInTime(DateTime? value, {DateTime? now}) {
  if (value == null) return 'Check-in date & time is required.';
  final reference = now ?? DateTime.now();
  if (value.isAfter(reference.add(const Duration(hours: maxFutureHours)))) {
    return 'Check-in time cannot be more than $maxFutureHours hours in the future.';
  }
  if (value.year < 2000) return 'Check-in time is too far in the past.';
  return null;
}

/// Blank ([value] null) is valid — it means "still staying".
String? validateCheckOutTime(DateTime? value, DateTime? checkInValue, {DateTime? now}) {
  if (value == null) return null;
  final reference = now ?? DateTime.now();
  if (value.isAfter(reference.add(const Duration(hours: maxFutureHours)))) {
    return 'Check-out time cannot be more than $maxFutureHours hours in the future.';
  }
  if (checkInValue != null && value.isBefore(checkInValue)) {
    return 'Check-out cannot be before check-in.';
  }
  return null;
}

String? validateGuestName(String value) => validateName(value);

/// Optional — blank is valid.
String? validateGuestPhone(String value) {
  if (value.trim().isEmpty) return null;
  return validatePhone(value);
}

final _pinPattern = RegExp(r'^\d{4,8}$');

String? validatePin(String pin) {
  if (!_pinPattern.hasMatch(pin)) return 'PIN must be 4-8 digits.';
  return null;
}

enum CustomerField {
  name,
  address,
  phone,
  roomNumber,
  numberOfPersons,
  amount,
  checkIn,
  checkOut,
  idFront,
  idBack,
}

typedef FieldErrors<T> = Map<T, String>;

bool hasErrors(FieldErrors errors) => errors.values.any((v) => v.isNotEmpty);

/// Raw values coming out of the check-in / edit form widget, before
/// normalisation into a [CustomerInput]-shaped payload for the service layer.
class CustomerFormValues {
  final String name;
  final String address;
  final String phone;
  final String roomNumber;
  final String numberOfPersonsText;
  final String amountText;
  final DateTime? checkInAt;
  final DateTime? checkOutAt;

  const CustomerFormValues({
    required this.name,
    required this.address,
    required this.phone,
    required this.roomNumber,
    required this.numberOfPersonsText,
    required this.amountText,
    required this.checkInAt,
    required this.checkOutAt,
  });

  CustomerFormValues copyWith({
    String? name,
    String? address,
    String? phone,
    String? roomNumber,
    String? numberOfPersonsText,
    String? amountText,
    DateTime? checkInAt,
    bool clearCheckInAt = false,
    DateTime? checkOutAt,
    bool clearCheckOutAt = false,
  }) =>
      CustomerFormValues(
        name: name ?? this.name,
        address: address ?? this.address,
        phone: phone ?? this.phone,
        roomNumber: roomNumber ?? this.roomNumber,
        numberOfPersonsText: numberOfPersonsText ?? this.numberOfPersonsText,
        amountText: amountText ?? this.amountText,
        checkInAt: clearCheckInAt ? null : (checkInAt ?? this.checkInAt),
        checkOutAt: clearCheckOutAt ? null : (checkOutAt ?? this.checkOutAt),
      );
}

FieldErrors<CustomerField> validateCustomerForm(
  CustomerFormValues values, {
  required bool hasIdFront,
  required bool hasIdBack,
  bool canEditCheckOut = false,
}) {
  final errors = <CustomerField, String>{};
  void put(CustomerField field, String? error) {
    if (error != null) errors[field] = error;
  }

  put(CustomerField.name, validateName(values.name));
  put(CustomerField.address, validateAddress(values.address));
  put(CustomerField.phone, validatePhone(values.phone));
  put(CustomerField.roomNumber, validateRoomNumber(values.roomNumber));
  put(CustomerField.numberOfPersons, validatePersons(values.numberOfPersonsText));
  put(CustomerField.amount, validateAmount(values.amountText));
  put(CustomerField.checkIn, validateCheckInTime(values.checkInAt));
  if (canEditCheckOut) {
    put(CustomerField.checkOut, validateCheckOutTime(values.checkOutAt, values.checkInAt));
  }
  if (!hasIdFront) put(CustomerField.idFront, 'Front ID photo is required.');
  if (!hasIdBack) put(CustomerField.idBack, 'Back ID photo is required.');
  return errors;
}

/// Trims/collapses whitespace and parses numeric fields into a service-layer
/// payload. Assumes [values] already passed [validateCustomerForm].
class NormalisedCustomerForm {
  final String name;
  final String address;
  final String phone;
  final String roomNumber;
  final int numberOfPersons;
  final int? amountMinor;
  final DateTime? checkInAt;
  final DateTime? checkOutAt;

  const NormalisedCustomerForm({
    required this.name,
    required this.address,
    required this.phone,
    required this.roomNumber,
    required this.numberOfPersons,
    required this.amountMinor,
    required this.checkInAt,
    required this.checkOutAt,
  });
}

NormalisedCustomerForm normaliseCustomerInput(CustomerFormValues values) {
  final name = values.name.trim().replaceAll(RegExp(r'\s+'), ' ');
  final parsedAmount = parseAmountToMinor(values.amountText);
  return NormalisedCustomerForm(
    name: name,
    address: values.address.trim(),
    phone: values.phone.trim(),
    roomNumber: values.roomNumber.trim(),
    numberOfPersons: int.tryParse(values.numberOfPersonsText.trim()) ?? 1,
    amountMinor: parsedAmount.valid ? parsedAmount.minor : null,
    checkInAt: values.checkInAt,
    checkOutAt: values.checkOutAt,
  );
}

class GuestFormValues {
  final String name;
  final String phone;
  const GuestFormValues({required this.name, required this.phone});
}

FieldErrors<CustomerField> validateGuestForm(GuestFormValues values) {
  final errors = <CustomerField, String>{};
  final nameError = validateGuestName(values.name);
  if (nameError != null) errors[CustomerField.name] = nameError;
  final phoneError = validateGuestPhone(values.phone);
  if (phoneError != null) errors[CustomerField.phone] = phoneError;
  return errors;
}

class NormalisedGuestForm {
  final String name;
  final String? phone;
  const NormalisedGuestForm({required this.name, required this.phone});
}

NormalisedGuestForm normaliseGuestInput(GuestFormValues values) {
  final name = values.name.trim().replaceAll(RegExp(r'\s+'), ' ');
  final phone = values.phone.trim();
  return NormalisedGuestForm(name: name, phone: phone.isEmpty ? null : phone);
}
